crates/op-rbuilder/src/gas_limiter/mod.rs

```diff
@@ -0,0 +1,122 @@
+
+pub mod args;
+pub mod error;
+mod metrics;
+
+#[derive(Clone)]
+pub struct AddressGasLimiter {
+    config: GasLimiterArgs,
+    // We don't need an Arc<Mutex<_>> here, we can get away with RefCell, but
+    // the reth PayloadBuilder trait needs this to be Send + Sync
+    address_buckets: Arc<Mutex<HashMap<Address, TokenBucket>>>,
+    metrics: GasLimiterMetrics,
+}
+
+impl std::fmt::Debug for AddressGasLimiter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AddressGasLimiter")
+            .field("config", &self.config)
+            .field("address_buckets", &self.address_buckets)
+            .field("metrics", &self.metrics)
+            .finish()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TokenBucket {
+    capacity: u64,
+    available: u64,
+}
+
+impl AddressGasLimiter {
+    pub fn try_new(config: GasLimiterArgs) -> Option<Self> {
+        if !config.enabled {
+            return None;
+        }
+
+        Some(Self {
+            config,
+            address_buckets: Default::default(),
+            metrics: Default::default(),
+        })
+    }
+
+    fn consume_gas_inner(
+        &self,
+        address: Address,
+        gas_requested: u64,
+    ) -> Result<bool, GasLimitError> {
+        let mut created_new_bucket = false;
+
+        let mut address_buckets = self.address_buckets.lock().unwrap();
+        let bucket = address_buckets
+            .entry(address)
+            // if we don't find a bucket we need to initialize a new one
+            .or_insert_with(|| {
+                created_new_bucket = true;
+                TokenBucket::new(self.config.max_gas_per_address)
+            });
+
+        if gas_requested > bucket.available {
+            return Err(GasLimitError::AddressLimitExceeded {
+                address,
+                requested: gas_requested,
+                available: bucket.available,
+            });
+        }
+
+        bucket.available -= gas_requested;
+
+        Ok(created_new_bucket)
+    }
+
+    /// Check if there's enough gas for this address and consume it. Returns
+    /// Ok(()) if there's enough otherwise returns an error.
+    pub fn consume_gas(&self, address: Address, gas_requested: u64) -> Result<(), GasLimitError> {
+        let start = Instant::now();
+        let result = self.consume_gas_inner(address, gas_requested);
+
+        self.metrics.record_gas_check(&result, start.elapsed());
+
+        result.map(|_| ())
+    }
+
+    /// Should be called upon each new block. Refills buckets/Garbage collection
+    fn refresh_inner(&self) -> usize {
+        let mut address_buckets = self.address_buckets.lock().unwrap();
+        let active_addresses = address_buckets.len();
+
+        address_buckets.iter_mut().for_each(|(_, bucket)| {
+            bucket.available += self.config.refill_rate_per_block;
+        });
+        address_buckets.retain(|_, bucket| bucket.available <= bucket.capacity);
```

Comment:
I think it's super bad for performance, if we created the bucket - let's keep it, otherwise we would always allocate memory 
Would make sense to remove bucket only if it wasn't used for 100 blocks (for example)